assembly/src/compile.rs

```diff
@@ -0,0 +1,332 @@
+    /// or deriving one from `kind`.
+    pub path: Option<LibraryPath>,
+}
+
+impl Default for Options {
+    fn default() -> Self {
+        Self {
+            kind: ModuleKind::Executable,
+            namespace: None,
+            path: None,
+        }
+    }
+}
+impl Options {
+    /// Configure a set of [Options] to compile a [Module] with the given `kind` and `path`.
+    ///
+    /// This is primarily useful when compiling a module from source code that has no meaningful
+    /// [LibraryPath] associated with it, such as when compiling from a `str`. This will override
+    /// the default name derived from the given [ModuleKind].
+    pub fn new<P, E>(kind: ModuleKind, path: P) -> Result<Self, E>
+    where
+        P: TryInto<LibraryPath, Error = E>,
+    {
+        let path = path.try_into()?;
+        Ok(Self {
+            kind,
+            namespace: None,
+            path: Some(path),
+        })
+    }
+
+    /// Get the default [Options] for compiling a library module.
+    pub fn for_library() -> Self {
+        Self {
+            kind: ModuleKind::Library,
+            ..Default::default()
+        }
+    }
+
+    /// Get the default [Options] for compiling a kernel module.
+    pub fn for_kernel() -> Self {
+        Self {
+            kind: ModuleKind::Kernel,
+            ..Default::default()
+        }
+    }
+}
+
+/// This error occurs when attempting to use a [Module] in a context which is invalid for that
+/// module.
+///
+/// For example, using an executable module as a library.
+#[derive(Debug, thiserror::Error, Diagnostic)]
+#[error("compilation failed: expected a {expected} module, but got a {actual} module")]
+#[diagnostic()]
+pub struct ModuleKindMismatchError {
+    expected: ModuleKind,
+    actual: ModuleKind,
+}
+
+// COMPILE TRAIT
+// ================================================================================================
+
+/// This trait is meant to be implemented by any type that can be compiled to a [Module],
+/// to allow methods which expect a [Module] to accept things like:
+///
+/// * A [Module] which was previously compiled or read from a [crate::Library].
+/// * A string representing the source code of a [Module].
+/// * A path to a file containing the source code of a [Module].
+/// * A vector of [Form]s comprising the contents of a [Module].
+pub trait Compile: Sized {
+    /// Compile (or convert) `self` into an executable [Module].
+    ///
+    /// See [Compile::compile_with_opts] for more details.
+    #[inline]
+    fn compile(self) -> Result<Box<Module>, Report> {
+        self.compile_with_opts(Options::default())
+    }
+
+    /// Compile (or convert) `self` into a [Module] using the provided `options`.
+    ///
+    /// Returns a [SyntaxError] if compilation fails due to a parsing or semantic analysis error,
+    /// or if the module provided is of the wrong kind (e.g. we expected a library module but got
```

Comment:
I think the reference here to `SyntaxError` is incorrect.

Reply:
Yeah, we're using `Report` now for these APIs, good catch.